enhancements/agent-installer/automated-workflow-kubeconfig.md

```diff
@@ -0,0 +1,253 @@
+replaces: N/A
+superseded-by: N/A
+---
+
+# Pre-generating the admin Kubeconfig in the Agent-based installer
+
+## Summary
+
+We will generate the admin kubeconfig file in the agent installer at the same
+time as the agent ISO. The CA keys used to sign the client certificate in the
+kubeconfig will be added to the ISO and eventually installed into the cluster
+so that the credentials in the admin kubeconfig will be accepted.
+
+## Motivation
+
+In the hosted assisted service, the admin kubeconfig is available for download
+from the web UI only after assisted-service has run openshift-installer to
+create the bootstrap ignition (i.e. after cluster installation has started).
+Since the hosted service is always available, this config can be downloaded at
+any time before the certificates in it expire.
+
+For the agent-based installer, doing the same would mean the kubeconfig was
+only available for a brief window of time, since the assisted-service is
+ephemeral. It runs on node 0 of the cluster and disappears after bootstrapping
+when that node is rebooted to become a cluster member.
+
+Furthermore, the kubeconfig file needs to be available to the installer in
+order for it to be used to monitor the cluster status in the `openshift-install
+agent wait-for install-complete` command.
+
+### User Stories
+
+* As a cluster creator, I want to have a reliable, non-timing dependent way to
+  retrieve the admin kubeconfig for a new cluster created with the agent-based
+  installer.
+
+* As a cluster creator, I want to use the `agent wait-for install-complete`
+  command to wait for the cluster to complete, with the preceding steps in the
+  automated workflow limited to only running the `agent create image` command
+  and booting the ISO on the hosts.
+
+### Goals
+
+* Generate a working admin kubeconfig as part of the `openshift-install agent
+  create image` command.
+
+### Non-Goals
+
+* Make the admin kubeconfig available earlier in the hosted assisted UI.
+
+## Proposal
+
+The admin kubeconfig will be generated by the `agent create image` command
+using an Asset based on the existing `kubeconfig.AdminClient` one. This will
+differ only in that it will get the cluster API address from the agent
+cluster-manifests, rather than the install-config, because the install-config
+is optional in the agent-based installer flow.
+
+The actual ignition files used for bootstrapping the cluster are not created by
+the agent-based installer binary, but rather by the assisted-service running
+openshift-installer again. In order to ensure that the cluster that is
+installed accepts the credentials in the kubeconfig, we must supply all assets
+that match the following rules:
+
+If we define an 'implicated' asset as one that:
+
+* Contains some additional data that cannot be recreated from its dependencies alone (unlike, e.g. the CABundle assets); and
+* Can be written to and loaded from disk; and
```

Comment:
Are there any assets that cannot be recreated and also cannot be written to and loaded from disk?

Reply:
Yes, in fact all of the interesting ones cannot be loaded from disk and that's exactly the point where this enhancement proposes to intervene.